File: api/routers/engagement.py
"""Story 6.5 — POST /api/v1/engagement: 웹 클라이언트발 engagement 이벤트 수집.

수집 대상은 open·read_through·impression뿐이다. decision은 서버 정본(decisions 라우터가 로깅)이라
클라이언트발을 금지한다(Pydantic Literal에서 제외 → 잘못된 타입은 422). variant도 클라이언트가 못
정한다(impression variant 정본은 recommender/서버, D2/D5) — 요청 스키마에 variant 필드 없음.

AD-13(API 계약): Bearer JWT(get_current_user 재사용)·/api/v1/·{data,error} 봉투.
AD-5(safe degradation): 각 이벤트는 best-effort insert — 존재하지 않는 signal_id는 FK 위반으로
조용히 스킵(accepted 카운트에서 제외)되고, 어떤 실패도 UX를 막지 않는다.
"""
from typing import Annotated, Literal

from fastapi import APIRouter, Depends
from pydantic import BaseModel

from core.schemas import APIResponse
from core.supabase import get_supabase
from middleware.auth import get_current_user
from pipeline.logger import pipeline_log

router = APIRouter(prefix="/engagement", tags=["engagement"])


class EngagementEvent(BaseModel):
    signal_id: str
    # 클라이언트발 허용 타입 — decision 제외(서버 정본). 잘못된 타입은 Pydantic이 422로 거절.
    event_type: Literal["impression", "open", "read_through"]
    daily_brief_id: str | None = None
    metadata: dict | None = None


class EngagementBatch(BaseModel):
    events: list[EngagementEvent]


@router.post("", response_model=APIResponse)
def collect_engagement(
    body: EngagementBatch,
    user_id: Annotated[str, Depends(get_current_user)],
) -> APIResponse:
    """배치 engagement 이벤트를 best-effort로 수집. accepted 카운트 반환(개별 실패는 전체를 막지 않음)."""
    client = get_supabase()

    accepted = 0
    for event in body.events:
        row = {
            "user_id": user_id,
            "signal_id": event.signal_id,
            "event_type": event.event_type,
        }
        if event.daily_brief_id:
            row["daily_brief_id"] = event.daily_brief_id
        if event.metadata is not None:
            row["metadata"] = event.metadata
        try:
            client.table("engagement_events").insert(row).execute()
            accepted += 1
        except Exception as e:  # noqa: BLE001 — best-effort: 개별 실패 스킵, UX 무영향(AD-5)
            # 예: 존재하지 않는 signal_id → FK 위반. accepted 증가 없이 경고만(AD-12).
            pipeline_log(
                stage="engagement",
                brief_date="",
                user_count=0,
                level="warning",
                event="engagement_log_failed",
                event_type=event.event_type,
                signal_id=event.signal_id,
                error=str(e)[:200],
            )

    return APIResponse(data={"accepted": accepted})

File: web/src/lib/engagement.ts
"use client";

/**
 * Story 6.5 — 웹 engagement 계측 헬퍼 (fire-and-forget).
 *
 * 프론트로 치면 Google Analytics의 track() 같은 것: 사용자 반응(open·read_through)을 백엔드에
 * 남긴다. 단, **화면을 절대 느리게 하면 안 된다** — 그래서 호출부는 await 하지 않고(fire-and-forget),
 * 여기서 세션 조회·fetch를 비동기로 던지고 실패는 전부 삼킨다. 렌더/네비게이션을 막지 않는다.
 *
 * decision·impression은 서버 정본이라 여기서 보내지 않는다(엔드포인트도 decision을 거절).
 * impression 타입을 허용 목록에 둔 건 향후 클라 뷰포트 계측 확장 여지를 위한 것으로, 지금 웹은
 * open·read_through만 전송한다(D2/D3).
 */
import { API_BASE_URL } from "@/lib/api-config";
import { getAccessToken } from "@/lib/api";

type EngagementEventType = "impression" | "open" | "read_through";

export interface EngagementEvent {
  signal_id: string;
  event_type: EngagementEventType;
  daily_brief_id?: string;
  metadata?: Record<string, unknown>;
}

export function trackEngagement(events: EngagementEvent[]): void {
  if (events.length === 0) return;
  // void + 즉시실행 async: 호출부는 반환을 기다리지 않는다(fire-and-forget).
  void (async () => {
    try {
      const token = await getAccessToken();
      if (!token) return; // 비로그인/세션만료 → 조용히 스킵(계측은 부수효과)

      await fetch(`${API_BASE_URL}/api/v1/engagement`, {
        method: "POST",
        headers: {
          "Content-Type": "application/json",
          Authorization: `Bearer ${token}`,
        },
        body: JSON.stringify({ events }),
        keepalive: true, // 페이지 이탈 중에도 전송 시도(read_through 직후 네비게이션 대비)
      });
    } catch {
      // fire-and-forget: 모든 실패 무시 (UX 무영향)
    }
  })();
}
